src/Components/WebAssembly/Server/src/DebugProxyLauncher.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    internal static class DebugProxyLauncher
+    {
+        private static readonly object LaunchLock = new object();
+        private static readonly TimeSpan DebugProxyLaunchTimeout = TimeSpan.FromSeconds(10);
+        private static Task<string> LaunchedDebugProxyUrl;
+
+        public static Task<string> EnsureLaunchedAndGetUrl(IServiceProvider serviceProvider)
+        {
+            lock (LaunchLock)
+            {
+                if (LaunchedDebugProxyUrl == null)
+                {
+                    LaunchedDebugProxyUrl = LaunchAndGetUrl(serviceProvider);
+                }
+
+                return LaunchedDebugProxyUrl;
+            }
+        }
+
+        private static async Task<string> LaunchAndGetUrl(IServiceProvider serviceProvider)
+        {
+            var tcs = new TaskCompletionSource<string>();
+
+            var environment = serviceProvider.GetRequiredService<IWebHostEnvironment>();
+            var executablePath = LocateDebugProxyExecutable(environment);
+            var muxerPath = DotNetMuxer.MuxerPathOrDefault();
+            var ownerPid = Process.GetCurrentProcess().Id;
+            var processStartInfo = new ProcessStartInfo
+            {
+                FileName = muxerPath,
+                Arguments = $"exec \"{executablePath}\" --owner-pid {ownerPid}",
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+            };
+            RemoveUnwantedEnvironmentVariables(processStartInfo.Environment);
+
+            var debugProxyProcess = Process.Start(processStartInfo);
+            var nowListeningRegex = new Regex("^\\s*Now listening on: (.*)$");
```

Comment:
Prior art: 

* https://github.com/dotnet/aspnetcore/blob/87a92e52c8b4bb7cb75ff78d53d641b1d34f8775/src/Hosting/Server.IntegrationTesting/src/Deployers/SelfHostDeployer.cs#L22
* https://github.com/dotnet/aspnetcore/blob/87a92e52c8b4bb7cb75ff78d53d641b1d34f8775/src/Hosting/Server.IntegrationTesting/src/Deployers/SelfHostDeployer.cs#L147-L161

Reply:
It's probably useful to wait for the app start message the way tests do this, since requests may start failing with 500s if the app isn't ready.

Reply:
Good tips. Done.